### PR TITLE
nixos/ssh: add crypto options

### PR DESCRIPTION
**Pull request by Izorkin**
###### Motivation for this change
Add custom options - PreferredAuthentications, KexAlgorithms, Ciphers, MACs.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

---
This change is [Reviewable](https://reviewable.io/reviews/nixos/nixpkgs/41798)

### CODE REVIEW

**Comment by grahamc** [comment]
Is there a nice benefit to using the structured form here, compared to the extraConfig?

**Comment by Izorkin** [comment]
It seems to me that it looks better. Also, the user can see the description here https://nixos.org/nixos/options.html#programs.ssh. and apply it in your config

**Comment by oxij** [comment]
:+1: on crypto options (since ssh is about crypto), I would even go as far as to turn all the crypto stuff we hardcode now into options.

:-1: on everything else (bloat, `man ssh` has a ton more useful options, should we add them too?).
While I agree that those `TCPKeepAlive` and etc options are nice, if I were you I would add them and maybe others as examples to `extraConfig` option instead.

Also, I feel like the usual convention on defaults with values that are not to be overridden by nix is to set them to `null`s not to valid values of the target type like `[]` and `3`.

Also, links between options should use "<option>" tag, not "see below". Options can be sorted pretty arbitrarily in the manual.

**Comment by Izorkin** [comment]
@oxij variant with kexAlgorithms set to null not worked, error - 
`error: The option value `programs.ssh.kexAlgorithms' in `/home/test/src_nix/nixpkgs/nixos/modules/programs/ssh.nix' is not of type `list of strings'.`
I do not plan to add all the options )

**Comment by oxij** [comment]
> type error

The type needs to be `types.nullOr (types.listOf types.str)` for that.

> I don't plan ...

But what is your criteria for adding something from `man ssh` to the nixos module? Usefulness at this moment?

NixOS is a fairly paranoid distro so we have a lot of crypto options already and adding some more wouldn't be out of style (and unhardcoding hardcoded ones would even be useful).

On the other hand, adding service bloat for some arbitrary set of options feels counterproductive to me.

I feel like what you really want here is pieces of `man ssh` documentation in `man configuration.nix` (aka Ted Nelson's Xanadu-style hypertext) which, indeed, is very useful. What is needed here is another generic hypertext mechanism like `relatedPackages`, not copy-paste bloat.

**Comment by qolii** [comment]
> But what is your criteria for adding something

Honest question: what are ever the criteria? Is there a guideline? It seems like every module in the whole OS is subject to an arbitrary decision on how much of the configuration space to support "properly". Most modules of significant complexity seem not to support it all.

I've always seen deferring of module options to an `extraConfig` to be a necessary (due to manpower), but ultimately undesirable, practice, and thought that the more of the configuration space is supported, the better. Especially for something like `sshd`, where these things are quite stable; once the module code is in place, it should keep working and hopefully not need much maintenance.

What are the downsides? Worry about maintenance? Or does it bog down Hydra or something too?

**Comment by oxij** [comment]
> policy

I'm not aware of any official policy, but I usually try to do this:

- things that don't have default values get their options,
- things overridden from their defaults by NixOS get options,
- nothing else gets an option unless it is used in generation from a
  declarative configuration in some way or is hard to specify with `extraConfig`.

> downsides

- Code bloat (KISS!).
- Maintenance of sync of descriptions between the mans of tool(section) and configuration.nix(5).
- Maintenance of default values overridden by the thing.
- Longer configuration.nix(5) manual, longer build system times.

Let me remind you that if you really want those options there in the manual you can define them and generate `extraConfig` from them and `ssh.enable` in _any_ nix file required from your configuration. I.e. you don't need to push those changes into NixOS core to use them. Personally, I do that a lot.

**Comment by coretemp** [comment]
If you write a script which translates sshd options into NixOS options, I think that would be something more people would welcome (at least, I would), but just adding code duplication (the ssh people already wrote that code) doesn't solve anything. 

It's not as if NixOS would continue to support module options if the underlying tool would have removed it. As such for options for which NixOS cannot guarantee its continued support, manually created options don't make sense.

**Comment by aanderse** [comment]
@Izorkin considering a few of these options have been implemented now and it seems unlikely the others will can we close this PR?

**Comment by Izorkin** [comment]
@aanderse, which options have been implemented? I've checked the code, and I don't see any changes to implement the options.

**Comment by aanderse** [comment]
@Izorkin `kexAlgorithms` and `ciphers` appear on https://nixos.org/nixos/options.html#services.openssh.

**Comment by Izorkin** [comment]
This optins in openssh server. Such options are available for the openssh client.
https://man.openbsd.org/sshd_config#KexAlgorithms
https://man.openbsd.org/ssh_config#KexAlgorithms

**Comment by aanderse** [comment]
@Izorkin ah sorry I only skimmed over this PR and didn't notice it was for the client software. Do you have any interest to rework this to include only the crypto options as @oxij mentioned?

**Comment by Izorkin** [comment]
@aanderse this variant normall?
Or preferredAuthentications remove?

**Comment by aanderse** [comment]
I'm not sure if @oxij just meant `kexAlgorithms` and `ciphers` but that's what I assume.

**Comment by Izorkin** [comment]
preferredAuthentications removed.

**Comment by Izorkin** [comment]
Example ssh config after this PR:
```
  programs.ssh = {
    ...
    macs = [ "hmac-sha2-512-etm@openssh.com" "hmac-sha2-512" ];
    ciphers = [ "chacha20-poly1305@openssh.com" ];
    kexAlgorithms = [ "curve25519-sha256@libssh.org" ];
    ...
  };
  
  services.openssh = {
    ...
    macs = [ "hmac-sha2-512-etm@openssh.com" ];
    ciphers = [ "chacha20-poly1305@openssh.com" ];
    kexAlgorithms = [ "curve25519-sha256@libssh.org" ];
    ...
  };
```

**Review by jonringer — APPROVED** [review]
diff LGTM
brings client ssh into feature parity with the openssh daemon

cc @Infinisil for final look-over

File: nixos/modules/programs/ssh.nix [hunk 1]
```diff
@@ -199,6 +199,33 @@ in
         '';
       };
 
+      kexAlgorithms = mkOption {
+        type = types.nullOr (types.listOf types.str);
+        default = null;
+        example = [ "curve25519-sha256@libssh.org" "diffie-hellman-group-exchange-sha256" ];
+        description = ''
+          Specifies the available KEX (Key Exchange) algorithms. Multiple algorithms must be comma-separated.
```
Comment (Infinisil) [inline nixos/modules/programs/ssh.nix:207]: The "Multiple algorithms must be comma-separated" part isn't relevant for the option and can be removed. Same for the other options.

**Review by infinisil — CHANGES_REQUESTED** [review]

File: nixos/modules/programs/ssh.nix (hunk 1 above)
Comment (Izorkin) [reply nixos/modules/programs/ssh.nix]: Fixed, thanks.

**Review by infinisil — APPROVED** [review]

**Pull request merged**